File: riskviewer-backend/src/main/java/com/satyam/riskviewer_backend/service/AdvancedFinancialModelsService.java
package com.satyam.riskviewer_backend.service;

import org.springframework.stereotype.Service;
import java.util.*;

@Service
public class AdvancedFinancialModelsService {

    // Black-Scholes Option Pricing
    public Map<String, Object> calculateBlackScholes(
            double spotPrice, double strikePrice, double timeToExpiry,
            double riskFreeRate, double volatility, double dividendYield) {
        
        double d1 = calculateD1(spotPrice, strikePrice, timeToExpiry, riskFreeRate, volatility, dividendYield);
        double d2 = d1 - volatility * Math.sqrt(timeToExpiry);
        
        double callPrice = spotPrice * Math.exp(-dividendYield * timeToExpiry) * normalCDF(d1)
                         - strikePrice * Math.exp(-riskFreeRate * timeToExpiry) * normalCDF(d2);
        
        double putPrice = strikePrice * Math.exp(-riskFreeRate * timeToExpiry) * normalCDF(-d2)
                        - spotPrice * Math.exp(-dividendYield * timeToExpiry) * normalCDF(-d1);

        Map<String, Object> greeks = calculateGreeks(spotPrice, strikePrice, timeToExpiry, 
                                                   riskFreeRate, volatility, dividendYield, d1, d2);

        Map<String, Object> result = new HashMap<>();
        result.put("callPrice", callPrice);
        result.put("putPrice", putPrice);
        result.put("greeks", greeks);
        
        return result;
    }

    private double calculateD1(double S, double K, double T, double r, double sigma, double q) {
        return (Math.log(S / K) + (r - q + 0.5 * sigma * sigma) * T) / (sigma * Math.sqrt(T));
    }

    private Map<String, Object> calculateGreeks(double S, double K, double T, double r, 
                                              double sigma, double q, double d1, double d2) {
        double sqrtT = Math.sqrt(T);
        double nd1 = normalPDF(d1);
        double Nd1 = normalCDF(d1);
        double Nd2 = normalCDF(d2);
        
        // Delta (price sensitivity)
        double callDelta = Math.exp(-q * T) * Nd1;
        
        // Gamma (delta sensitivity)
        double gamma = Math.exp(-q * T) * nd1 / (S * sigma * sqrtT);
        
        // Theta (time decay)
        double callTheta = (-S * nd1 * sigma * Math.exp(-q * T) / (2 * sqrtT)
                          - r * K * Math.exp(-r * T) * Nd2
                          + q * S * Math.exp(-q * T) * Nd1) / 365;
        
        // Vega (volatility sensitivity)
        double vega = S * Math.exp(-q * T) * nd1 * sqrtT / 100;
        
        // Rho (interest rate sensitivity)
        double callRho = K * T * Math.exp(-r * T) * Nd2 / 100;
        
        Map<String, Object> greeks = new HashMap<>();
        greeks.put("delta", callDelta);
        greeks.put("gamma", gamma);
        greeks.put("theta", callTheta);
        greeks.put("vega", vega);
        greeks.put("rho", callRho);
        
        return greeks;
    }

    // Credit Risk Assessment
    public Map<String, Object> calculateCreditRisk(double creditScore, double debtToEquity,
                                                  double currentRatio, double interestCoverage,
                                                  double industryRisk, double exposureAmount) {
        
        double pd = calculateProbabilityOfDefault(creditScore, debtToEquity, currentRatio, 
                                                interestCoverage, industryRisk);
        double lgd = calculateLossGivenDefault(debtToEquity, currentRatio);
        double ead = exposureAmount;
        
        double expectedLoss = pd * lgd * ead;
        double unexpectedLoss = Math.sqrt(pd * (1 - pd)) * lgd * ead;
        double creditVaR = normalInverse(0.99) * unexpectedLoss + expectedLoss;
        String creditRating = getCreditRating(pd);

        Map<String, Object> result = new HashMap<>();
        result.put("probabilityOfDefault", pd);
        result.put("lossGivenDefault", lgd);
        result.put("exposureAtDefault", ead);
        result.put("expectedLoss", expectedLoss);
        result.put("unexpectedLoss", unexpectedLoss);
        result.put("creditVaR", creditVaR);
        result.put("creditRating", creditRating);
        
        return result;
    }

    private double calculateProbabilityOfDefault(double creditScore, double debtToEquity,
                                               double currentRatio, double interestCoverage,
                                               double industryRisk) {
        // Logistic regression model for PD calculation
        double scoreWeight = -0.003;
        double debtWeight = 0.15;
        double liquidityWeight = -0.08;
        double coverageWeight = -0.02;
        double industryWeight = 0.05;
        
        double logit = 2.5 +
                      scoreWeight * creditScore +
                      debtWeight * debtToEquity +
                      liquidityWeight * currentRatio +
                      coverageWeight * interestCoverage +
                      industryWeight * industryRisk;
        
        return Math.max(0.001, Math.min(0.999, 1 / (1 + Math.exp(-logit))));
    }

    private double calculateLossGivenDefault(double debtToEquity, double currentRatio) {
        double baseLGD = 0.45; // 45% base loss given default
        double debtAdjustment = Math.min(0.3, debtToEquity * 0.1);
        double liquidityAdjustment = Math.max(-0.2, -currentRatio * 0.05);
        
        return Math.max(0.1, Math.min(0.9, baseLGD + debtAdjustment + liquidityAdjustment));
    }

    private String getCreditRating(double pd) {
        if (pd < 0.001) return "AAA";
        if (pd < 0.002) return "AA+";
        if (pd < 0.005) return "AA";
        if (pd < 0.01) return "AA-";
        if (pd < 0.02) return "A+";
        if (pd < 0.05) return "A";
        if (pd < 0.1) return "A-";
        if (pd < 0.15) return "BBB+";
        if (pd < 0.25) return "BBB";
        if (pd < 0.4) return "BBB-";
        if (pd < 0.6) return "BB+";
        if (pd < 0.8) return "BB";
        return "B";
    }

    // Stress Testing Scenarios
    public List<Map<String, Object>> performStressTests(double portfolioValue) {
        List<Map<String, Object>> results = new ArrayList<>();
        
        results.add(createStressTestResult("Market Crash (2008-style)",
                "Severe market downturn with 40% equity decline, 20% bond decline",
                portfolioValue, -0.32, 150, 42, 24));
        
        results.add(createStressTestResult("Interest Rate Shock",
                "300bp sudden interest rate increase affecting bonds and equity",
                portfolioValue, -0.23, 80, 25, 18));
        
        results.add(createStressTestResult("Pandemic Crisis (COVID-19 style)",
                "Global pandemic causing economic shutdown and market volatility",
                portfolioValue, -0.35, 200, 38, 12));
        
        results.add(createStressTestResult("Inflation Spike",
                "Persistent high inflation (8%+) eroding real returns",
                portfolioValue, -0.18, 60, 22, 36));
        
        results.add(createStressTestResult("Geopolitical Crisis",
                "Major geopolitical conflict affecting global markets",
                portfolioValue, -0.25, 120, 28, 15));
        
        results.add(createStressTestResult("Liquidity Crisis",
                "Severe liquidity crunch with credit markets freezing",
                portfolioValue, -0.30, 180, 35, 20));
        
        return results;
    }

    private Map<String, Object> createStressTestResult(String scenario, String description,
                                                     double portfolioValue, double percentageChange,
                                                     int var95Change, int maxDrawdown, int recovery) {
        Map<String, Object> result = new HashMap<>();
        double valueChange = portfolioValue * percentageChange;
        
        result.put("scenario", scenario);
        result.put("description", description);
        result.put("portfolioValue", portfolioValue + valueChange);
        result.put("valueChange", valueChange);
        result.put("percentageChange", percentageChange * 100);
        result.put("var95Change", var95Change);
        result.put("maxDrawdown", maxDrawdown);
        result.put("recovery", recovery);
        
        return result;
    }

    // Liquidity Risk Assessment
    public Map<String, Object> calculateLiquidityMetrics(double cashAndEquivalents,
                                                        double liquidSecurities, double totalAssets,
                                                        double shortTermLiabilities, double totalLiabilities) {
        
        double liquidityCoverageRatio = (cashAndEquivalents + liquidSecurities) / (shortTermLiabilities * 0.3);
        double netStableFundingRatio = (totalAssets - shortTermLiabilities) / totalAssets;
        double liquidityBuffer = cashAndEquivalents + liquidSecurities;
        double cashRatio = cashAndEquivalents / shortTermLiabilities;
        double quickRatio = (cashAndEquivalents + liquidSecurities) / shortTermLiabilities;
        
        String liquidityRisk = "LOW";
        if (liquidityCoverageRatio < 1.0 || quickRatio < 1.0) {
            liquidityRisk = "HIGH";
        } else if (liquidityCoverageRatio < 1.5 || quickRatio < 1.5) {
            liquidityRisk = "MEDIUM";
        }

        Map<String, Object> result = new HashMap<>();
        result.put("liquidityCoverageRatio", liquidityCoverageRatio);
        result.put("netStableFundingRatio", netStableFundingRatio);
        result.put("liquidityBuffer", liquidityBuffer);
        result.put("cashRatio", cashRatio);
        result.put("quickRatio", quickRatio);
        result.put("liquidityRisk", liquidityRisk);
        
        return result;
    }

    // Utility functions for statistical calculations
    private double normalCDF(double x) {
        return (1.0 + erf(x / Math.sqrt(2.0))) / 2.0;
    }

    private double normalPDF(double x) {
        return Math.exp(-0.5 * x * x) / Math.sqrt(2 * Math.PI);
    }

    private double normalInverse(double p) {
        // Approximation of inverse normal distribution
        if (p <= 0) return Double.NEGATIVE_INFINITY;
        if (p >= 1) return Double.POSITIVE_INFINITY;
        
        // Beasley-Springer-Moro approximation
        double a0 = 2.50662823884;
        double a1 = -18.61500062529;
        double a2 = 41.39119773534;
        double a3 = -25.44106049637;
        double b1 = -8.47351093090;
        double b2 = 23.08336743743;
        double b3 = -21.06224101826;
        double b4 = 3.13082909833;
        double c0 = 0.3374754822726147;
        double c1 = 0.9761690190917186;
        double c2 = 0.1607979714918209;
        double c3 = 0.0276438810333863;
        double c4 = 0.0038405729373609;
        double c5 = 0.0003951896511919;
        double c6 = 0.0000321767881768;
        double c7 = 0.0000002888167364;
        double c8 = 0.0000003960315187;
        
        double x = p - 0.5;
        
        if (Math.abs(x) < 0.42) {
            double r = x * x;
            return x * (((a3 * r + a2) * r + a1) * r + a0) /
                   ((((b4 * r + b3) * r + b2) * r + b1) * r + 1);
        }
        
        double r = p < 0.5 ? p : 1 - p;
        r = Math.sqrt(-Math.log(r));
        
        double result = (((((((c8 * r + c7) * r + c6) * r + c5) * r + c4) * r + c3) * r + c2) * r + c1) * r + c0;
        
        return p < 0.5 ? -result : result;
    }

    private double erf(double x) {
        // Approximation of error function
        double a1 = 0.254829592;
        double a2 = -0.284496736;
        double a3 = 1.421413741;
        double a4 = -1.453152027;
        double a5 = 1.061405429;
        double p = 0.3275911;

        int sign = x >= 0 ? 1 : -1;
        x = Math.abs(x);

        double t = 1.0 / (1.0 + p * x);
        double y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * Math.exp(-x * x);

        return sign * y;
    }
}

File: riskviewer-frontend/src/app/services/advanced-financial-models.service.ts
import { Injectable } from '@angular/core';

export interface BlackScholesResult {
  callPrice: number;
  putPrice: number;
  greeks: {
    delta: number;
    gamma: number;
    theta: number;
    vega: number;
    rho: number;
  };
}

export interface CreditRiskMetrics {
  probabilityOfDefault: number;
  lossGivenDefault: number;
  exposureAtDefault: number;
  expectedLoss: number;
  unexpectedLoss: number;
  creditVaR: number;
  creditRating: string;
}

export interface StressTestResult {
  scenario: string;
  description: string;
  portfolioValue: number;
  valueChange: number;
  percentageChange: number;
  var95Change: number;
  maxDrawdown: number;
  recovery: number;
}

export interface LiquidityMetrics {
  liquidityCoverageRatio: number;
  netStableFundingRatio: number;
  liquidityBuffer: number;
  cashRatio: number;
  quickRatio: number;
  liquidityRisk: string;
}

@Injectable({
  providedIn: 'root'
})
export class AdvancedFinancialModelsService {

  constructor() { }

  // Black-Scholes Option Pricing Model
  calculateBlackScholes(
    spotPrice: number,
    strikePrice: number,
    timeToExpiry: number,
    riskFreeRate: number,
    volatility: number,
    dividendYield: number = 0
  ): BlackScholesResult {
    
    const d1 = this.calculateD1(spotPrice, strikePrice, timeToExpiry, riskFreeRate, volatility, dividendYield);
    const d2 = d1 - volatility * Math.sqrt(timeToExpiry);
    
    const callPrice = spotPrice * Math.exp(-dividendYield * timeToExpiry) * this.normalCDF(d1) 
                    - strikePrice * Math.exp(-riskFreeRate * timeToExpiry) * this.normalCDF(d2);
    
    const putPrice = strikePrice * Math.exp(-riskFreeRate * timeToExpiry) * this.normalCDF(-d2) 
                   - spotPrice * Math.exp(-dividendYield * timeToExpiry) * this.normalCDF(-d1);

    // Calculate Greeks
    const greeks = this.calculateGreeks(spotPrice, strikePrice, timeToExpiry, riskFreeRate, volatility, dividendYield, d1, d2);

    return {
      callPrice,
      putPrice,
      greeks
    };
  }

  private calculateD1(S: number, K: number, T: number, r: number, sigma: number, q: number): number {
    return (Math.log(S / K) + (r - q + 0.5 * sigma * sigma) * T) / (sigma * Math.sqrt(T));
  }

  private calculateGreeks(S: number, K: number, T: number, r: number, sigma: number, q: number, d1: number, d2: number) {
    const sqrtT = Math.sqrt(T);
    const nd1 = this.normalPDF(d1);
    const Nd1 = this.normalCDF(d1);
    const Nd2 = this.normalCDF(d2);
    
    // Delta (price sensitivity)
    const callDelta = Math.exp(-q * T) * Nd1;
    const putDelta = -Math.exp(-q * T) * this.normalCDF(-d1);
    
    // Gamma (delta sensitivity)
    const gamma = Math.exp(-q * T) * nd1 / (S * sigma * sqrtT);
    
    // Theta (time decay)
    const callTheta = (-S * nd1 * sigma * Math.exp(-q * T) / (2 * sqrtT) 
                     - r * K * Math.exp(-r * T) * Nd2 
                     + q * S * Math.exp(-q * T) * Nd1) / 365;
    
    // Vega (volatility sensitivity)
    const vega = S * Math.exp(-q * T) * nd1 * sqrtT / 100;
    
    // Rho (interest rate sensitivity)
    const callRho = K * T * Math.exp(-r * T) * Nd2 / 100;
    
    return {
      delta: callDelta,
      gamma: gamma,
      theta: callTheta,
      vega: vega,
      rho: callRho
    };
  }

  // Credit Risk Assessment
  calculateCreditRisk(
    creditScore: number,
    debtToEquity: number,
    currentRatio: number,
    interestCoverage: number,
    industryRisk: number,
    exposureAmount: number
  ): CreditRiskMetrics {
    
    // Simplified credit risk model based on financial ratios
    let pd = this.calculateProbabilityOfDefault(creditScore, debtToEquity, currentRatio, interestCoverage, industryRisk);
    let lgd = this.calculateLossGivenDefault(debtToEquity, currentRatio);
    let ead = exposureAmount;
    
    const expectedLoss = pd * lgd * ead;
    const unexpectedLoss = Math.sqrt(pd * (1 - pd)) * lgd * ead;
    const creditVaR = this.normalInverse(0.99) * unexpectedLoss + expectedLoss;
    const creditRating = this.getCreditRating(pd);

    return {
      probabilityOfDefault: pd,
      lossGivenDefault: lgd,
      exposureAtDefault: ead,
      expectedLoss,
      unexpectedLoss,
      creditVaR,
      creditRating
    };
  }

  private calculateProbabilityOfDefault(creditScore: number, debtToEquity: number, currentRatio: number, interestCoverage: number, industryRisk: number): number {
    // Logistic regression model for PD calculation
    const scoreWeight = -0.003;
    const debtWeight = 0.15;
    const liquidityWeight = -0.08;
    const coverageWeight = -0.02;
    const industryWeight = 0.05;
    
    const logit = 2.5 + 
                  scoreWeight * creditScore + 
                  debtWeight * debtToEquity + 
                  liquidityWeight * currentRatio + 
                  coverageWeight * interestCoverage + 
                  industryWeight * industryRisk;
    
    return Math.max(0.001, Math.min(0.999, 1 / (1 + Math.exp(-logit))));
  }

  private calculateLossGivenDefault(debtToEquity: number, currentRatio: number): number {
    // LGD based on collateral and seniority
    const baseLGD = 0.45; // 45% base loss given default
    const debtAdjustment = Math.min(0.3, debtToEquity * 0.1);
    const liquidityAdjustment = Math.max(-0.2, -currentRatio * 0.05);
    
    return Math.max(0.1, Math.min(0.9, baseLGD + debtAdjustment + liquidityAdjustment));
  }

  private getCreditRating(pd: number): string {
    if (pd < 0.001) return 'AAA';
    if (pd < 0.002) return 'AA+';
    if (pd < 0.005) return 'AA';
    if (pd < 0.01) return 'AA-';
    if (pd < 0.02) return 'A+';
    if (pd < 0.05) return 'A';
    if (pd < 0.1) return 'A-';
    if (pd < 0.15) return 'BBB+';
    if (pd < 0.25) return 'BBB';
    if (pd < 0.4) return 'BBB-';
    if (pd < 0.6) return 'BB+';
    if (pd < 0.8) return 'BB';
    return 'B';
  }

  // Stress Testing Scenarios
  performStressTests(portfolioValue: number, positions: any[]): StressTestResult[] {
    return [
      this.marketCrashScenario(portfolioValue, positions),
      this.interestRateShockScenario(portfolioValue, positions),
      this.covidPandemicScenario(portfolioValue, positions),
      this.inflationSpike(portfolioValue, positions),
      this.geopoliticalCrisis(portfolioValue, positions),
      this.liquidityCrisis(portfolioValue, positions),
      this.sectorConcentrationRisk(portfolioValue, positions)
    ];
  }

  private marketCrashScenario(portfolioValue: number, positions: any[]): StressTestResult {
    // 2008-style market crash: -40% equity, -20% bonds, +50% vol
    const equityWeight = 0.7; // Assume 70% equity
    const bondWeight = 0.3;   // Assume 30% bonds
    
    const valueChange = portfolioValue * (equityWeight * -0.40 + bondWeight * -0.20);
    const newValue = portfolioValue + valueChange;
    
    return {
      scenario: 'Market Crash (2008-style)',
      description: 'Severe market downturn with 40% equity decline, 20% bond decline, volatility spike',
      portfolioValue: newValue,
      valueChange,
      percentageChange: (valueChange / portfolioValue) * 100,
      var95Change: 150, // VaR increases by 150%
      maxDrawdown: 42,
      recovery: 24 // months to recover
    };
  }

  private interestRateShockScenario(portfolioValue: number, positions: any[]): StressTestResult {
    // Sudden 300bp rate increase
    const bondWeight = 0.3;
    const duration = 7; // Average portfolio duration
    const rateShock = 0.03; // 3% rate increase
    
    const bondImpact = -bondWeight * duration * rateShock;
    const equityImpact = -0.15; // Equity down 15% due to higher discount rates
    
    const valueChange = portfolioValue * (bondImpact + equityImpact * 0.7);
    
    return {
      scenario: 'Interest Rate Shock',
      description: '300bp sudden interest rate increase affecting bonds and equity valuations',
      portfolioValue: portfolioValue + valueChange,
      valueChange,
      percentageChange: (valueChange / portfolioValue) * 100,
      var95Change: 80,
      maxDrawdown: 25,
      recovery: 18
    };
  }

  private covidPandemicScenario(portfolioValue: number, positions: any[]): StressTestResult {
    // COVID-19 style scenario
    const valueChange = portfolioValue * -0.35; // 35% portfolio decline
    
    return {
      scenario: 'Pandemic Crisis (COVID-19 style)',
      description: 'Global pandemic causing economic shutdown and market volatility',
      portfolioValue: portfolioValue + valueChange,
      valueChange,
      percentageChange: -35,
      var95Change: 200,
      maxDrawdown: 38,
      recovery: 12
    };
  }

  private inflationSpike(portfolioValue: number, positions: any[]): StressTestResult {
    // High inflation scenario
    const valueChange = portfolioValue * -0.18; // Real value erosion
    
    return {
      scenario: 'Inflation Spike',
      description: 'Persistent high inflation (8%+) eroding real returns',
      portfolioValue: portfolioValue + valueChange,
      valueChange,
      percentageChange: -18,
      var95Change: 60,
      maxDrawdown: 22,
      recovery: 36
    };
  }

  private geopoliticalCrisis(portfolioValue: number, positions: any[]): StressTestResult {
    const valueChange = portfolioValue * -0.25;
    
    return {
      scenario: 'Geopolitical Crisis',
      description: 'Major geopolitical conflict affecting global markets',
      portfolioValue: portfolioValue + valueChange,
      valueChange,
      percentageChange: -25,
      var95Change: 120,
      maxDrawdown: 28,
      recovery: 15
    };
  }

  private liquidityCrisis(portfolioValue: number, positions: any[]): StressTestResult {
    const valueChange = portfolioValue * -0.30;
    
    return {
      scenario: 'Liquidity Crisis',
      description: 'Severe liquidity crunch with credit markets freezing',
      portfolioValue: portfolioValue + valueChange,
      valueChange,
      percentageChange: -30,
      var95Change: 180,
      maxDrawdown: 35,
      recovery: 20
    };
  }

  private sectorConcentrationRisk(portfolioValue: number, positions: any[]): StressTestResult {
    const valueChange = portfolioValue * -0.22;
    
    return {
      scenario: 'Sector Concentration Risk',
      description: 'Major sector collapse affecting concentrated positions',
      portfolioValue: portfolioValue + valueChange,
      valueChange,
      percentageChange: -22,
      var95Change: 90,
      maxDrawdown: 26,
      recovery: 18
    };
  }

  // Liquidity Risk Assessment
  calculateLiquidityMetrics(
    cashAndEquivalents: number,
    liquidSecurities: number,
    totalAssets: number,
    shortTermLiabilities: number,
    totalLiabilities: number,
    deposits: number
  ): LiquidityMetrics {
    
    const liquidityCoverageRatio = (cashAndEquivalents + liquidSecurities) / (shortTermLiabilities * 0.3);
    const netStableFundingRatio = (totalAssets - shortTermLiabilities) / totalAssets;
    const liquidityBuffer = cashAndEquivalents + liquidSecurities;
    const cashRatio = cashAndEquivalents / shortTermLiabilities;
    const quickRatio = (cashAndEquivalents + liquidSecurities) / shortTermLiabilities;
    
    let liquidityRisk = 'LOW';
    if (liquidityCoverageRatio < 1.0 || quickRatio < 1.0) liquidityRisk = 'HIGH';
    else if (liquidityCoverageRatio < 1.5 || quickRatio < 1.5) liquidityRisk = 'MEDIUM';

    return {
      liquidityCoverageRatio,
      netStableFundingRatio,
      liquidityBuffer,
      cashRatio,
      quickRatio,
      liquidityRisk
    };
  }

  // Utility functions for statistical calculations
  private normalCDF(x: number): number {
    return (1.0 + this.erf(x / Math.sqrt(2.0))) / 2.0;
  }

  private normalPDF(x: number): number {
    return Math.exp(-0.5 * x * x) / Math.sqrt(2 * Math.PI);
  }

  private normalInverse(p: number): number {
    // Approximation of inverse normal distribution
    if (p <= 0) return -Infinity;
    if (p >= 1) return Infinity;
    
    const a1 = -3.969683028665376e+01;
    const a2 = 2.209460984245205e+02;
    const a3 = -2.759285104469687e+02;
    const a4 = 1.383577518672690e+02;
    const a5 = -3.066479806614716e+01;
    const a6 = 2.506628277459239e+00;

    const b1 = -5.447609879822406e+01;
    const b2 = 1.615858368580409e+02;
    const b3 = -1.556989798598866e+02;
    const b4 = 6.680131188771972e+01;
    const b5 = -1.328068155288572e+01;

    const c1 = -7.784894002430293e-03;
    const c2 = -3.223964580411365e-01;
    const c3 = -2.400758277161838e+00;
    const c4 = -2.549732539343734e+00;
    const c5 = 4.374664141464968e+00;
    const c6 = 2.938163982698783e+00;

    const d1 = 7.784695709041462e-03;
    const d2 = 3.224671290700398e-01;
    const d3 = 2.445134137142996e+00;
    const d4 = 3.754408661907416e+00;

    let q, t, u;

    if (p < 0.02425) {
      q = Math.sqrt(-2 * Math.log(p));
      return (((((c1 * q + c2) * q + c3) * q + c4) * q + c5) * q + c6) /
             ((((d1 * q + d2) * q + d3) * q + d4) * q + 1);
    } else if (p < 0.97575) {
      q = p - 0.5;
      t = q * q;
      return (((((a1 * t + a2) * t + a3) * t + a4) * t + a5) * t + a6) * q /
             (((((b1 * t + b2) * t + b3) * t + b4) * t + b5) * t + 1);
    } else {
      q = Math.sqrt(-2 * Math.log(1 - p));
      return -(((((c1 * q + c2) * q + c3) * q + c4) * q + c5) * q + c6) /
              ((((d1 * q + d2) * q + d3) * q + d4) * q + 1);
    }
  }

  private erf(x: number): number {
    // Approximation of error function
    const a1 =  0.254829592;
    const a2 = -0.284496736;
    const a3 =  1.421413741;
    const a4 = -1.453152027;
    const a5 =  1.061405429;
    const p  =  0.3275911;

    const sign = x >= 0 ? 1 : -1;
    x = Math.abs(x);

    const t = 1.0 / (1.0 + p * x);
    const y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * Math.exp(-x * x);

    return sign * y;
  }
}
